rust/numaflow-core/src/pipeline.rs

```diff
@@ -41,6 +42,24 @@ pub(crate) async fn start_forwarder(
         pipeline::VertexType::Source(source) => {
             info!("Starting source forwarder");
 
+            // Create a new cancellation if the source is Serving.
+            // The new cancellation token gets cancelled only after the user specified
+            // request timeout + 30 seconds. The `cln_token` gets cancelled when SIGTERM is received.
+            if let SourceType::Serving(ref serving_config) = source.source_config.source_type {
+                let serving_cln_token = CancellationToken::new();
+                tokio::spawn({
+                    let req_timeout = serving_config.drain_timeout_secs + 30;
+                    let cln_token = cln_token.clone();
+                    let serving_cln_token = serving_cln_token.clone();
+                    async move {
+                        cln_token.cancelled().await;
+                        tokio::time::sleep(Duration::from_secs(req_timeout)).await;
+                        serving_cln_token.cancel();
+                    }
```

Comment:
we shouldn't do a blind sleep, we can return early if there are no requests-in-flight.

Reply:
The [handle](https://docs.rs/axum-server/latest/axum_server/struct.Handle.html) provides a method to know current connection count. We may have to find a way to pass this token to `serving` crate and cancel it from there.